presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java

```diff
@@ -336,28 +336,41 @@ protected Object visitIsNotNullPredicate(IsNotNullPredicate node, Object context
         protected Object visitSearchedCaseExpression(SearchedCaseExpression node, Object context)
         {
             Expression resultClause = node.getDefaultValue().orElse(null);
+
+            ImmutableList.Builder<WhenClause> builder = ImmutableList.builder();
             for (WhenClause whenClause : node.getWhenClauses()) {
                 Object value = process(whenClause.getOperand(), context);
-                if (value instanceof Expression) {
-                    // TODO: optimize this case
-                    return node;
-                }
-
-                if (Boolean.TRUE.equals(value)) {
-                    resultClause = whenClause.getResult();
-                    break;
+                if (value != null) {
```

Comment:
This check shouldn't be necessary. If the value is null, it will fail both conditions of the `if` block below.